From: My Omarchy contributors
Subject: [PATCH] cocoa: present the VM as My Omarchy on macOS

Applies after startergo/homebrew-qemu-virgl-kosmickrisp v1.0.27.

QEMU owns the visible Cocoa window, but the launcher remains the product's
single Dock application and management entry point. Publish the product process
name, keep QEMU as an accessory application, load the signed app icon from the
enclosing bundle, and use the product name in QEMU's Cocoa window and menu.

diff --git a/ui/cocoa.m b/ui/cocoa.m
--- a/ui/cocoa.m
+++ b/ui/cocoa.m
@@ -1197,7 +1197,7 @@
     COCOA_DEBUG("QemuCocoaView: grabMouse\n");
 
     if (qemu_name)
-        [[self window] setTitle:[NSString stringWithFormat:@"QEMU %s - (Press  " UC_CTRL_KEY " " UC_ALT_KEY " G  to release Mouse)", qemu_name]];
+        [[self window] setTitle:[NSString stringWithFormat:@"%s - (Press  " UC_CTRL_KEY " " UC_ALT_KEY " G  to release Mouse)", qemu_name]];
     else
         [[self window] setTitle:@"QEMU - (Press  " UC_CTRL_KEY " " UC_ALT_KEY " G  to release Mouse)"];
     [self hideCursor];
@@ -1210,7 +1210,7 @@
     COCOA_DEBUG("QemuCocoaView: ungrabMouse\n");
 
     if (qemu_name)
-        [[self window] setTitle:[NSString stringWithFormat:@"QEMU %s", qemu_name]];
+        [[self window] setTitle:[NSString stringWithUTF8String:qemu_name]];
     else
         [[self window] setTitle:@"QEMU"];
     [self unhideCursor];
@@ -1327,9 +1327,12 @@
         }
         [window setAcceptsMouseMovedEvents:YES];
         [window setCollectionBehavior:NSWindowCollectionBehaviorFullScreenPrimary];
-        [window setTitle:qemu_name ? [NSString stringWithFormat:@"QEMU %s", qemu_name] : @"QEMU"];
+        [window setTitle:qemu_name
+            ? [NSString stringWithUTF8String:qemu_name]
+            : @"QEMU"];
         [window setContentView:cocoaView];
-        [window makeKeyAndOrderFront:self];
+        [window orderFrontRegardless];
+        [window makeKeyWindow];
         [window center];
 
         /* Careful: Cocoa has a concept of a main window and a key window. */
@@ -1651,7 +1653,7 @@
 {
     NSAlert *alert = [NSAlert new];
     [alert autorelease];
-    [alert setMessageText: @"Are you sure you want to quit QEMU?"];
+    [alert setMessageText: @"Are you sure you want to quit My Omarchy?"];
     [alert addButtonWithTitle: @"Cancel"];
     [alert addButtonWithTitle: @"Quit"];
     if([alert runModal] == NSAlertSecondButtonReturn) {
@@ -1665,20 +1667,16 @@
 - (IBAction) do_about_menu_item: (id) sender
 {
     NSAutoreleasePool *pool = [[NSAutoreleasePool alloc] init];
-    char *icon_path_c = get_relocated_path(CONFIG_QEMU_ICONDIR "/hicolor/512x512/apps/qemu.png");
-    NSString *icon_path = [NSString stringWithUTF8String:icon_path_c];
-    g_free(icon_path_c);
-    NSImage *icon = [[NSImage alloc] initWithContentsOfFile:icon_path];
     NSString *version = @"QEMU emulator version " QEMU_FULL_VERSION;
     NSString *copyright = @QEMU_COPYRIGHT;
-    NSDictionary *options;
+    NSImage *icon = [NSApp applicationIconImage];
+    NSDictionary *options = nil;
     if (icon) {
         options = @{
             NSAboutPanelOptionApplicationIcon : icon,
             NSAboutPanelOptionApplicationVersion : version,
             @"Copyright" : copyright,
         };
-        [icon release];
     } else {
         options = @{
             NSAboutPanelOptionApplicationVersion : version,
@@ -1745,17 +1743,17 @@
 
     // Application menu
     menu = [[NSMenu alloc] initWithTitle:@""];
-    [menu addItemWithTitle:@"About QEMU" action:@selector(do_about_menu_item:) keyEquivalent:@""]; // About QEMU
+    [menu addItemWithTitle:@"About My Omarchy" action:@selector(do_about_menu_item:) keyEquivalent:@""];
     [menu addItem:[NSMenuItem separatorItem]]; //Separator
     menuItem = [menu addItemWithTitle:@"Services" action:nil keyEquivalent:@""];
     [menuItem setSubmenu:[NSApp servicesMenu]];
     [menu addItem:[NSMenuItem separatorItem]];
-    [menu addItemWithTitle:@"Hide QEMU" action:@selector(hide:) keyEquivalent:@"h"]; //Hide QEMU
+    [menu addItemWithTitle:@"Hide My Omarchy" action:@selector(hide:) keyEquivalent:@"h"];
     menuItem = (NSMenuItem *)[menu addItemWithTitle:@"Hide Others" action:@selector(hideOtherApplications:) keyEquivalent:@"h"]; // Hide Others
     [menuItem setKeyEquivalentModifierMask:(NSEventModifierFlagOption|NSEventModifierFlagCommand)];
     [menu addItemWithTitle:@"Show All" action:@selector(unhideAllApplications:) keyEquivalent:@""]; // Show All
     [menu addItem:[NSMenuItem separatorItem]]; //Separator
-    [menu addItemWithTitle:@"Quit QEMU" action:@selector(terminate:) keyEquivalent:@"q"];
+    [menu addItemWithTitle:@"Quit My Omarchy" action:@selector(terminate:) keyEquivalent:@"q"];
     menuItem = [[NSMenuItem alloc] initWithTitle:@"Apple" action:nil keyEquivalent:@""];
     [menuItem setSubmenu:menu];
     [[NSApp mainMenu] addItem:menuItem];
@@ -2421,7 +2419,31 @@
     assert(o->type == DISPLAY_TYPE_COCOA);
     if (o->has_gl && o->gl) {
         display_opengl = 1;
+    }
+}
+
+static void cocoa_set_product_icon(void)
+{
+    NSArray *arguments = [[NSProcessInfo processInfo] arguments];
+
+    if ([arguments count] == 0) {
+        return;
     }
+
+    NSString *executablePath =
+        [[arguments objectAtIndex:0] stringByStandardizingPath];
+    NSString *resourcesPath =
+        [[[executablePath stringByDeletingLastPathComponent]
+          stringByDeletingLastPathComponent]
+         stringByDeletingLastPathComponent];
+    NSString *iconPath =
+        [resourcesPath stringByAppendingPathComponent:@"MyOmarchy.icns"];
+    NSImage *icon = [[NSImage alloc] initWithContentsOfFile:iconPath];
+
+    if (icon) {
+        [NSApp setApplicationIconImage:icon];
+        [icon release];
+    }
 }
 
 static void cocoa_display_init(DisplayState *ds, DisplayOptions *opts)
@@ -2436,7 +2458,7 @@
-    ProcessSerialNumber psn = { 0, kCurrentProcess };
-    TransformProcessType(&psn, kProcessTransformToForegroundApplication);
-
+    [[NSProcessInfo processInfo] setProcessName:@"My Omarchy"];
     [QemuApplication sharedApplication];
+    [NSApp setActivationPolicy:NSApplicationActivationPolicyAccessory];
+    cocoa_set_product_icon();
 
     dcl.con = qemu_console_lookup_default();
     kbd = qkbd_state_init(dcl.con);
